ouroboros-network-api/src/Ouroboros/Network/BlockFetch/ConsensusInterface.hs

```diff
@@ -1,24 +1,35 @@
-{-# LANGUAGE DeriveFunctor #-}
-{-# LANGUAGE RankNTypes    #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveFunctor  #-}
+{-# LANGUAGE DeriveGeneric  #-}
+{-# LANGUAGE RankNTypes     #-}
 
 module Ouroboros.Network.BlockFetch.ConsensusInterface
   ( FetchMode (..)
+  , GenesisFetchMode (..)
   , BlockFetchConsensusInterface (..)
   , WhetherReceivingTentativeBlocks (..)
   , FromConsensus (..)
+  , ChainSelStarvation (..)
   ) where
 
 import Control.Monad.Class.MonadSTM
 import Control.Monad.Class.MonadTime (UTCTime)
+import Control.Monad.Class.MonadTime.SI (Time)
 
 import Data.Map.Strict (Map)
+import GHC.Generics (Generic)
 import GHC.Stack (HasCallStack)
+import NoThunks.Class (NoThunks)
 
 import Ouroboros.Network.AnchoredFragment (AnchoredFragment)
 import Ouroboros.Network.Block
 import Ouroboros.Network.SizeInBytes (SizeInBytes)
 
 
+-- | The fetch mode that the block fetch logic should use.
+data GenesisFetchMode = FetchModeGenesis | PraosFetchMode FetchMode
+  deriving (Eq, Show)
```

Comment:
FTR we talked about this in the Networking call yesterday:

 - When we integrate this into Consensus, we will make sure that when Genesis is enabled, [`readFetchModeDefault`](https://github.com/intersectmbo/ouroboros-consensus/blob/5a1409be0543592222d6b0ea4115bfa406c0e5bd/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/BlockFetch/ClientInterface.hs#L134-L166) will uses the current `LedgerStateJudgement` to decide which `FetchMode` to use (concretely, `FetchModeGenesis` when `TooOld`, and `PraosFetchMode FetchModeDeadline` when `YoungEnough`). (cc @crocodile-dentist)
 - However, when Genesis is disabled, we want to keep the current behavior (based on slot distance to the tip), so it makes sense to keep `FetchMode`/`GenesisFetchMode` for now.

Does that make sense?

Reply:
I think this will be confusing to use `FetchMode` and `LedgerStateJudgement` in this way.   Why not just directly use `LedgerStateJudgement`?  Can we use `FetchMode` in the legacy `block-fetch` implementation, but `LedgerStateJudgment` in the new one?  There's `ConsensusMode` available from `ouroboros-network-api` (which is passed through `Ouroboros.Network.Diffusion.P2P.ArugmentsExtra` from `consensus`) - this is the switch that is required.

Reply:
We could nicely abstract this using something like this:
```
data EffectiveFetchMode = BulkSyncMode | DeadlineMode

effectiveFetchMode :: ConsensusMode -> m FetchMode -> m LedgerStateJudgement -> m EffectiveFetchMode
effectiveFetchMode = ...
```

Reply:
So I think we are in agreement of what the semantics should be, i.e. when to use old/new BulkSync vs Deadline logic in Praos/Genesis mode:

 - In Praos mode: Use the existing `readFetchModeDefault` logic
 - In Genesis mode: Use the current `LedgerStateJudgement` to decide, ie `FetchModeGenesis` when `TooOld`, and `FetchModeDeadline` when `YoungEnough`.

The question now is where this logic should live.

 - It could live entirely in Consensus, such that the BlockFetch logic in Network still is only aware of the `FetchMode`.
 - It could be spread across Network and Consensus, such that Network makes the decision whether to use the `FetchMode` or the `LedgerStateJudgement`, basically your `effectiveFetchMode`.

I have no strong opinion here; the second option would require passing down more stuff to Network for BlockFetch, but we already pass down these things in other places, so it doesn't really increase the API surface between Network and Consensus.

Reply:
> So I think we are in agreement of what the semantics should be
Yes we are

> The question now is where this logic should live.

It would be nice to design something that's consistent, so it's easy which parts belong to genesis and which to praos. If it's just about the genesis block-fetch implementation it also could just directly use `LedgerStateJudgement` (it will only run if `GenesisMode` is on).

Reply:
I added the following function:
```haskell
-- | Construct 'readFetchMode' for 'BlockFetchConsensusInterface' by branching
-- on the 'ConsensusMode'.
mkReadFetchMode
  :: Functor m
  => ConsensusMode
  -> m LedgerStateJudgement
     -- ^ Used for 'GenesisMode'.
  -> m FetchMode
     -- ^ Used for 'PraosMode' for backwards compatibility.
  -> m GenesisFetchMode
mkReadFetchMode consensusMode getLedgerStateJudgement getFetchMode =
    case consensusMode of
      GenesisMode -> getLedgerStateJudgement <&> \case
        YoungEnough -> PraosFetchMode FetchModeDeadline
        TooOld      -> FetchModeGenesis
      PraosMode   -> PraosFetchMode <$> getFetchMode
```
and used it in Consensus when constructing the `BlockFetchConsensusInterface`.

This is similar to what was discussed [on Discord](https://discord.com/channels/1136727663583698984/1239889324745429122/1287699968802426882); I still kept `GenesisFetchMode` as it is more convenient to specify in e.g. tests and demos.